File: mac/VibeTunnel/Core/Services/SessionMonitor.swift
import Foundation
import Observation
import os.log
import UserNotifications

/// Server session information returned by the API.
///
/// Represents the current state of a terminal session running on the VibeTunnel server,
/// including its command, directory, process status, and activity information.
struct ServerSessionInfo: Codable {
    let id: String
    let name: String
    let command: [String]
    let workingDir: String
    let status: String
    let exitCode: Int?
    let startedAt: String
    let pid: Int?
    let initialCols: Int?
    let initialRows: Int?
    let lastClearOffset: Int?
    let version: String?
    let gitRepoPath: String?
    let gitBranch: String?
    let gitAheadCount: Int?
    let gitBehindCount: Int?
    let gitHasChanges: Bool?
    let gitIsWorktree: Bool?
    let gitMainRepoPath: String?

    // Additional fields from Session (not SessionInfo)
    let lastModified: String
    let active: Bool?
    let activityStatus: ActivityStatus?
    let source: String?
    let remoteId: String?
    let remoteName: String?
    let remoteUrl: String?
    let attachedViaVT: Bool?

    var isRunning: Bool {
        status == "running"
    }
}

/// Activity status for a session.
///
/// Tracks whether a session is actively being used and provides
/// application-specific status information when available.
struct ActivityStatus: Codable {
    let isActive: Bool
    let specificStatus: SpecificStatus?
}

/// App-specific status information.
///
/// Provides detailed status information for specific applications running
/// within a terminal session, such as Claude's current working state.
struct SpecificStatus: Codable {
    let app: String
    let status: String
}

/// Lightweight session monitor that fetches terminal sessions on-demand.
///
/// Manages the collection of active terminal sessions by periodically polling
/// the server API and caching results for efficient access. Provides real-time
/// session information to the UI with minimal network overhead.
@MainActor
@Observable
final class SessionMonitor {
    static let shared = SessionMonitor()

    /// Previous session states for exit detection
    private var previousSessions: [String: ServerSessionInfo] = [:]
    private var firstFetchDone = false

    /// Track last known activity state per session for Claude transition detection
    private var lastActivityState: [String: Bool] = [:]
    /// Sessions that have already triggered a "Your Turn" alert
    private var claudeIdleNotified: Set<String> = []

    /// Detect sessions that transitioned from running to not running
    static func detectEndedSessions(
        from old: [String: ServerSessionInfo],
        to new: [String: ServerSessionInfo]
    )
        -> [ServerSessionInfo]
    {
        old.compactMap { id, oldSession in
            if oldSession.isRunning,
               let updated = new[id], !updated.isRunning
            {
                return oldSession
            }
            return nil
        }
    }

    private(set) var sessions: [String: ServerSessionInfo] = [:]
    private(set) var lastError: Error?

    private var lastFetch: Date?
    private let cacheInterval: TimeInterval = 2.0
    private let serverManager = ServerManager.shared
    private let logger = Logger(subsystem: BundleIdentifiers.loggerSubsystem, category: "SessionMonitor")

    /// Reference to GitRepositoryMonitor for pre-caching
    weak var gitRepositoryMonitor: GitRepositoryMonitor?

    /// Timer for periodic refresh
    private var refreshTimer: Timer?

    private init() {
        // Start periodic refresh
        startPeriodicRefresh()
    }

    /// Set the local auth token for server requests
    func setLocalAuthToken(_ token: String?) {}

    /// Number of running sessions
    var sessionCount: Int {
        sessions.values.count { $0.isRunning }
    }

    /// Get all sessions, using cache if available
    func getSessions() async -> [String: ServerSessionInfo] {
        // Use cache if available and fresh
        if let lastFetch, Date().timeIntervalSince(lastFetch) < cacheInterval {
            return sessions
        }

        await fetchSessions()
        return sessions
    }

    /// Force refresh session data
    func refresh() async {
        lastFetch = nil
        await fetchSessions()
    }

    // MARK: - Private Methods

    private func fetchSessions() async {
        do {
            // Snapshot previous sessions for exit notifications
            let oldSessions = sessions

            let sessionsArray = try await serverManager.performRequest(
                endpoint: APIEndpoints.sessions,
                method: "GET",
                responseType: [ServerSessionInfo].self
            )

            // Convert to dictionary
            var sessionsDict: [String: ServerSessionInfo] = [:]
            for session in sessionsArray {
                sessionsDict[session.id] = session
            }

            self.sessions = sessionsDict
            self.lastError = nil

            // Notify for sessions that have just ended
            if firstFetchDone && UserDefaults.standard.bool(forKey: "showNotifications") {
                let ended = Self.detectEndedSessions(from: oldSessions, to: sessionsDict)
                for session in ended {
                    let id = session.id
                    let title = "Session Completed"
                    let displayName = session.name
                    let content = UNMutableNotificationContent()
                    content.title = title
                    content.body = displayName
                    content.sound = .default
                    let request = UNNotificationRequest(identifier: "session_\(id)", content: content, trigger: nil)
                    do {
                        try await UNUserNotificationCenter.current().add(request)
                    } catch {
                        self.logger
                            .error(
                                "Failed to deliver session notification: \(error.localizedDescription, privacy: .public)"
                            )
                    }
                }

                // Detect Claude "Your Turn" transitions
                await detectAndNotifyClaudeTurns(from: oldSessions, to: sessionsDict)
            }

            // Set firstFetchDone AFTER detecting ended sessions
            firstFetchDone = true
            self.lastFetch = Date()

            // Update WindowTracker
            WindowTracker.shared.updateFromSessions(sessionsArray)

            // Pre-cache Git data for all sessions (deduplicated by repository)
            if let gitMonitor = gitRepositoryMonitor {
                await preCacheGitRepositories(for: sessionsArray, using: gitMonitor)
            }
        } catch {
            // Only update error if it's not a simple connection error
            if !(error is URLError) {
                self.lastError = error
            }
            logger.error("Failed to fetch sessions: \(error, privacy: .public)")
            self.sessions = [:]
            self.lastFetch = Date() // Still update timestamp to avoid hammering
        }
    }

    /// Pre-cache Git repositories for sessions, deduplicating by repository root
    private func preCacheGitRepositories(for sessions: [ServerSessionInfo], using gitMonitor: GitRepositoryMonitor) async {
        // Track unique directories we need to check
        var uniqueDirectoriesToCheck = Set<String>()

        // First, collect all unique directories that don't have cached data
        for session in sessions {
            // Skip if we already have cached data for this exact path
            if gitMonitor.getCachedRepository(for: session.workingDir) != nil {
                continue
            }

            // Add this directory to check
            uniqueDirectoriesToCheck.insert(session.workingDir)

            // Smart detection: Also check common parent directories
            // This helps when multiple sessions are in subdirectories of the same project
            let pathComponents = session.workingDir.split(separator: "/").map(String.init)

            // Check if this looks like a project directory pattern
            // Common patterns: /Users/*/Projects/*, /Users/*/Development/*, etc.
            if pathComponents.count >= 4 {
                // Check if we're in a common development directory
                let commonDevPaths = ["Projects", "Development", "Developer", "Code", "Work", "Source"]

                for (index, component) in pathComponents.enumerated() {
                    if commonDevPaths.contains(component) && index < pathComponents.count - 1 {
                        // This might be a parent project directory
                        // Add the immediate child of the development directory
                        let potentialProjectPath = "/" + pathComponents[0...index + 1].joined(separator: "/")

                        // Only add if we don't have cached data for it
                        if gitMonitor.getCachedRepository(for: potentialProjectPath) == nil {
                            uniqueDirectoriesToCheck.insert(potentialProjectPath)
                        }
                    }
                }
            }
        }

        // Now check each unique directory only once
        for directory in uniqueDirectoriesToCheck {
            Task {
                // This will cache the data for immediate access later
                _ = await gitMonitor.findRepository(for: directory)
            }
        }

        logger
            .debug(
                "Pre-caching Git data for \(uniqueDirectoriesToCheck.count) unique directories (from \(sessions.count) sessions)"
            )
    }

    /// Start periodic refresh of sessions
    private func startPeriodicRefresh() {
        // Clean up any existing timer
        refreshTimer?.invalidate()

        // Create a new timer that fires every 3 seconds
        refreshTimer = Timer.scheduledTimer(withTimeInterval: 3.0, repeats: true) { [weak self] _ in
            Task { @MainActor [weak self] in
                await self?.refresh()
            }
        }
    }

    /// Detect and notify when Claude sessions transition from active to inactive ("Your Turn")
    private func detectAndNotifyClaudeTurns(
        from old: [String: ServerSessionInfo],
        to new: [String: ServerSessionInfo]
    )
        async
    {
        // Check if Claude notifications are enabled using ConfigManager
        let claudeNotificationsEnabled = ConfigManager.shared.notificationClaudeTurn
        guard claudeNotificationsEnabled else { return }

        for (id, newSession) in new {
            // Only process running sessions
            guard newSession.isRunning else { continue }

            // Check if this is a Claude session
            let isClaudeSession = newSession.activityStatus?.specificStatus?.app.lowercased()
                .contains("claude") ?? false ||
                newSession.command.joined(separator: " ").lowercased().contains("claude")

            guard isClaudeSession else { continue }

            // Get current activity state
            let currentActive = newSession.activityStatus?.isActive ?? false

            // Get previous activity state (from our tracking or old session data)
            let previousActive = lastActivityState[id] ?? (old[id]?.activityStatus?.isActive ?? false)

            // Reset when Claude speaks again
            if !previousActive && currentActive {
                claudeIdleNotified.remove(id)
            }

            // First active ➜ idle transition ⇒ alert
            let alreadyNotified = claudeIdleNotified.contains(id)
            if previousActive && !currentActive && !alreadyNotified {
                logger.info("🔔 Detected Claude transition to idle for session: \(id)")
                let sessionName = newSession.name ?? newSession.command.joined(separator: " ")
                
                // Create a claude-turn event for the notification
                let claudeTurnEvent = ServerEvent.claudeTurn(
                    sessionId: id,
                    sessionName: sessionName
                )
                await NotificationService.shared.sendNotification(for: claudeTurnEvent)
                claudeIdleNotified.insert(id)
            }

            // Update tracking *after* detection logic
            lastActivityState[id] = currentActive
        }

        // Clean up tracking for ended/closed sessions
        for id in lastActivityState.keys {
            if new[id] == nil || !(new[id]?.isRunning ?? false) {
                lastActivityState.removeValue(forKey: id)
                claudeIdleNotified.remove(id)
            }
        }
    }
}

File: web/src/server/services/session-monitor.ts
/**
 * SessionMonitor - Server-side monitoring of terminal sessions
 *
 * Replaces the Mac app's polling-based SessionMonitor with real-time
 * event detection directly from PTY streams. Tracks session states,
 * command execution, and Claude-specific activity transitions.
 */

import { EventEmitter } from 'events';
import { ServerEventType } from '../../shared/types.js';
import type { PtyManager } from '../pty/pty-manager.js';
import { createLogger } from '../utils/logger.js';

const logger = createLogger('session-monitor');

// Command tracking thresholds
const MIN_COMMAND_DURATION_MS = 3000; // Minimum duration for command completion notifications
const CLAUDE_IDLE_DEBOUNCE_MS = 2000; // Debounce period for Claude idle detection

export interface SessionState {
  id: string;
  name: string;
  command: string[];
  workingDir: string;
  status: 'running' | 'exited';
  isRunning: boolean;
  pid?: number;

  // Activity tracking
  activityStatus?: {
    isActive: boolean;
    lastActivity?: Date;
    specificStatus?: {
      app: string;
      status: string;
    };
  };

  // Command tracking
  commandStartTime?: Date;
  lastCommand?: string;
  lastExitCode?: number;

  // Claude-specific tracking
  isClaudeSession?: boolean;
  claudeActivityState?: 'active' | 'idle' | 'unknown';
}

export interface CommandFinishedEvent {
  sessionId: string;
  sessionName: string;
  command: string;
  duration: number;
  exitCode: number;
}

export interface ClaudeTurnEvent {
  sessionId: string;
  sessionName: string;
  message?: string;
}

export class SessionMonitor extends EventEmitter {
  private sessions = new Map<string, SessionState>();
  private claudeIdleNotified = new Set<string>();
  private lastActivityState = new Map<string, boolean>();
  private commandThresholdMs = MIN_COMMAND_DURATION_MS;
  private claudeIdleTimers = new Map<string, NodeJS.Timeout>();

  constructor(private ptyManager: PtyManager) {
    super();
    this.setupEventListeners();
    logger.info('SessionMonitor initialized');
  }

  private setupEventListeners() {
    // Listen for session lifecycle events
    this.ptyManager.on('sessionStarted', (sessionId: string, sessionName: string) => {
      this.handleSessionStarted(sessionId, sessionName);
    });

    this.ptyManager.on(
      'sessionExited',
      (sessionId: string, sessionName: string, exitCode?: number) => {
        this.handleSessionExited(sessionId, sessionName, exitCode);
      }
    );

    // Listen for command tracking events
    this.ptyManager.on('commandFinished', (data: CommandFinishedEvent) => {
      this.handleCommandFinished(data);
    });

    // Listen for Claude activity events (if available)
    this.ptyManager.on('claudeTurn', (sessionId: string, sessionName: string) => {
      this.handleClaudeTurn(sessionId, sessionName);
    });
  }

  /**
   * Update session state with activity information
   */
  public updateSessionActivity(sessionId: string, isActive: boolean, specificApp?: string) {
    const session = this.sessions.get(sessionId);
    if (!session) return;

    const previousActive = session.activityStatus?.isActive ?? false;

    // Update activity status
    session.activityStatus = {
      isActive,
      lastActivity: isActive ? new Date() : session.activityStatus?.lastActivity,
      specificStatus: specificApp
        ? {
            app: specificApp,
            status: isActive ? 'active' : 'idle',
          }
        : session.activityStatus?.specificStatus,
    };

    // Check if this is a Claude session
    if (this.isClaudeSession(session)) {
      this.trackClaudeActivity(sessionId, session, previousActive, isActive);
    }

    this.lastActivityState.set(sessionId, isActive);
  }

  /**
   * Track PTY output for activity detection and bell characters
   */
  public trackPtyOutput(sessionId: string, data: string) {
    const session = this.sessions.get(sessionId);
    if (!session) return;

    // Update last activity
    this.updateSessionActivity(sessionId, true);

    // Detect bell character
    if (data.includes('\x07')) {
      this.emit('bell', {
        sessionId,
        sessionName: session.name,
        timestamp: new Date().toISOString(),
      });
    }

    // Detect Claude-specific patterns in output
    if (this.isClaudeSession(session)) {
      this.detectClaudePatterns(sessionId, session, data);
    }
  }

  /**
   * Update command information for a session
   */
  public updateCommand(sessionId: string, command: string) {
    const session = this.sessions.get(sessionId);
    if (!session) return;

    session.lastCommand = command;
    session.commandStartTime = new Date();

    // Mark as active when a new command starts
    this.updateSessionActivity(sessionId, true);
  }

  /**
   * Handle command completion
   */
  public handleCommandCompletion(sessionId: string, exitCode: number) {
    const session = this.sessions.get(sessionId);
    if (!session || !session.commandStartTime || !session.lastCommand) return;

    const duration = Date.now() - session.commandStartTime.getTime();
    session.lastExitCode = exitCode;

    // Only emit event if command ran long enough
    if (duration >= this.commandThresholdMs) {
      const _event: CommandFinishedEvent = {
        sessionId,
        sessionName: session.name,
        command: session.lastCommand,
        duration,
        exitCode,
      };

      // Emit appropriate event based on exit code
      if (exitCode === 0) {
        this.emit('notification', {
          type: ServerEventType.CommandFinished,
          sessionId,
          sessionName: session.name,
          command: session.lastCommand,
          duration,
          exitCode,
        });
      } else {
        this.emit('notification', {
          type: ServerEventType.CommandError,
          sessionId,
          sessionName: session.name,
          command: session.lastCommand,
          duration,
          exitCode,
        });
      }
    }

    // Clear command tracking
    session.commandStartTime = undefined;
    session.lastCommand = undefined;
  }

  private handleSessionStarted(sessionId: string, sessionName: string) {
    // Get full session info from PtyManager
    const ptySession = this.ptyManager.getSession(sessionId);
    if (!ptySession) return;

    const state: SessionState = {
      id: sessionId,
      name: sessionName,
      command: ptySession.command || [],
      workingDir: ptySession.workingDir || process.cwd(),
      status: 'running',
      isRunning: true,
      pid: ptySession.pid,
      isClaudeSession: this.detectClaudeCommand(ptySession.command || []),
    };

    this.sessions.set(sessionId, state);
    logger.info(`Session started: ${sessionId} - ${sessionName}`);

    // Emit notification event
    this.emit('notification', {
      type: ServerEventType.SessionStart,
      sessionId,
      sessionName,
      timestamp: new Date().toISOString(),
    });
  }

  private handleSessionExited(sessionId: string, sessionName: string, exitCode?: number) {
    const session = this.sessions.get(sessionId);
    if (!session) return;

    session.status = 'exited';
    session.isRunning = false;

    logger.info(`Session exited: ${sessionId} - ${sessionName} (exit code: ${exitCode})`);

    // Clean up Claude tracking
    this.claudeIdleNotified.delete(sessionId);
    this.lastActivityState.delete(sessionId);
    if (this.claudeIdleTimers.has(sessionId)) {
      const timer = this.claudeIdleTimers.get(sessionId);
      if (timer) clearTimeout(timer);
      this.claudeIdleTimers.delete(sessionId);
    }

    // Emit notification event
    this.emit('notification', {
      type: ServerEventType.SessionExit,
      sessionId,
      sessionName,
      exitCode,
      timestamp: new Date().toISOString(),
    });

    // Remove session after a delay to allow final events to process
    setTimeout(() => {
      this.sessions.delete(sessionId);
    }, 5000);
  }

  private handleCommandFinished(data: CommandFinishedEvent) {
    // Forward to our handler which will emit the appropriate notification
    this.handleCommandCompletion(data.sessionId, data.exitCode);
  }

  private handleClaudeTurn(sessionId: string, _sessionName: string) {
    const session = this.sessions.get(sessionId);
    if (!session) return;

    // Mark Claude as idle
    this.updateSessionActivity(sessionId, false, 'claude');
  }

  private isClaudeSession(session: SessionState): boolean {
    return session.isClaudeSession ?? false;
  }

  private detectClaudeCommand(command: string[]): boolean {
    const commandStr = command.join(' ').toLowerCase();
    return commandStr.includes('claude');
  }

  private trackClaudeActivity(
    sessionId: string,
    session: SessionState,
    previousActive: boolean,
    currentActive: boolean
  ) {
    // Clear any existing idle timer
    if (this.claudeIdleTimers.has(sessionId)) {
      const timer = this.claudeIdleTimers.get(sessionId);
      if (timer) clearTimeout(timer);
      this.claudeIdleTimers.delete(sessionId);
    }

    // Claude went from active to potentially idle
    if (previousActive && !currentActive && !this.claudeIdleNotified.has(sessionId)) {
      // Set a debounce timer before declaring Claude idle
      const timer = setTimeout(() => {
        // Check if still idle
        const currentSession = this.sessions.get(sessionId);
        if (currentSession?.activityStatus && !currentSession.activityStatus.isActive) {
          logger.info(`🔔 Claude turn detected for session: ${sessionId}`);

          this.emit('notification', {
            type: ServerEventType.ClaudeTurn,
            sessionId,
            sessionName: session.name,
            message: 'Claude has finished responding',
            timestamp: new Date().toISOString(),
          });

          this.claudeIdleNotified.add(sessionId);
        }

        this.claudeIdleTimers.delete(sessionId);
      }, CLAUDE_IDLE_DEBOUNCE_MS);

      this.claudeIdleTimers.set(sessionId, timer);
    }

    // Claude became active again - reset notification flag
    if (!previousActive && currentActive) {
      this.claudeIdleNotified.delete(sessionId);
    }
  }

  private detectClaudePatterns(sessionId: string, _session: SessionState, data: string) {
    // Detect patterns that indicate Claude is working or has finished
    const workingPatterns = ['Thinking...', 'Analyzing', 'Working on', 'Let me'];

    const idlePatterns = [
      "I've completed",
      "I've finished",
      'Done!',
      "Here's",
      'The task is complete',
    ];

    // Check for working patterns
    for (const pattern of workingPatterns) {
      if (data.includes(pattern)) {
        this.updateSessionActivity(sessionId, true, 'claude');
        return;
      }
    }

    // Check for idle patterns
    for (const pattern of idlePatterns) {
      if (data.includes(pattern)) {
        // Delay marking as idle to allow for follow-up output
        setTimeout(() => {
          this.updateSessionActivity(sessionId, false, 'claude');
        }, 1000);
        return;
      }
    }
  }

  /**
   * Get all active sessions
   */
  public getActiveSessions(): SessionState[] {
    return Array.from(this.sessions.values()).filter((s) => s.isRunning);
  }

  /**
   * Get a specific session
   */
  public getSession(sessionId: string): SessionState | undefined {
    return this.sessions.get(sessionId);
  }

  /**
   * Initialize monitor with existing sessions
   */
  public async initialize() {
    // Get all existing sessions from PtyManager
    const existingSessions = await this.ptyManager.listSessions();

    for (const session of existingSessions) {
      if (session.status === 'running') {
        const state: SessionState = {
          id: session.id,
          name: session.name,
          command: session.command,
          workingDir: session.workingDir,
          status: 'running',
          isRunning: true,
          pid: session.pid,
          isClaudeSession: this.detectClaudeCommand(session.command),
        };

        this.sessions.set(session.id, state);
      }
    }

    logger.info(`Initialized with ${this.sessions.size} existing sessions`);
  }
}
